### PR TITLE
Modularize build: charset, options

### PR DESCRIPTION
**Pull request by regisd**
* Move Out in jflex.logging
* Move Options in jflex.option
* Create bazel target for jflex.core.unicode
* Move the charclass related code in jflex.core.unicode
* Introduce ILexScan interface to decoupe the unicode classes from the actual LexScan implementation (which is in jflex.core)
* Create jflex.core.OptionUtils for setting the complex options. setDefaults() must be called before the generation is invoked, and this is guarded with the encoding field.
* Reduce bazel target visibility
* Moved DEBUG const in Build class

### CODE REVIEW

File: testsuite/testcases/src/test/cases/java/java-2.input [hunk 1]
```diff
@@ -134,7 +134,7 @@ final public class NFA {
   public void addRegExp(int regExpNum) {
 
     if (Options.DEBUG)
-      Out.debug("Adding nfa for regexp "+regExpNum+" :"+Out.NL+regExps.getRegExp(regExpNum));
+      Out.debug("Adding nfa for regexp "+regExpNum+" :"+Env.NL+regExps.getRegExp(regExpNum));
```
Comment (lsf37) [inline testsuite/testcases/src/test/cases/java/java-2.input]: The search/replace was a bit too global here, this file should remain the same.

File: jflex/src/main/java/jflex/env/Env.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package jflex.env;
+
+public class Env {
+
+  private Env() {}
```
Comment (lsf37) [inline jflex/src/main/java/jflex/env/Env.java]: Codacy wants the constructor after the fields. Probably fair.

**Comment by regisd** [comment]
> Good idea in general, just minor things to fix up (one the input files in the testsuite happens to be java code from some old JFlex version that mentions Out -- that should remain as is)

This is WIP. I want to actually to do a more serious modularization. NL was just a small annoyance. I hope I'll get something.

**Comment by lsf37** [comment]
No worries.

**Comment by regisd** [comment]
Just pushed my latest update, this is basically moving
- Out in logging package
- char-classes related code in unicode package
- core depends on both logging and unicode packages

The problem is still with `Options` which remains in core. Some Out methods depend on Options. I can't have cyclic dependencies.

File: jflex/src/main/jflex/LexScan.flex [hunk 3]
```diff
@@ -13,9 +13,9 @@ import java_cup.runtime.Symbol;
 import java.io.File;
 import java.io.FileNotFoundException;
 import java.io.FileReader;
-import jflex.core.unicode.UnicodeProperties;
+import jflex.core.unicode.IntCharSet;import jflex.core.unicode.UnicodeProperties;
```
Comment (lsf37) [inline jflex/src/main/jflex/LexScan.flex]: typo

File: jflex/src/main/jflex/LexScan.flex [hunk 4]
```diff
@@ -13,9 +13,9 @@ import java_cup.runtime.Symbol;
...
 import jflex.l10n.ErrorMessages;
-import jflex.performance.Timer;
+import jflex.logging.Out;import jflex.performance.Timer;
```
Comment (lsf37) [inline jflex/src/main/jflex/LexScan.flex]: typo

**Comment by lsf37** [comment]
> Just pushed my latest update, this is basically moving
> 
> * Out in logging package
> * char-classes related code in unicode package
> * core depends on both logging and unicode packages

In principle no problem with that.

> The problem is still with `Options` which remains in core. Some Out methods depend on Options. I can't have cyclic dependencies.

Hm. This is tricky. I don't see how logging should not depend on runtime options. Maybe from the other direction. Maybe a separate module for Options and related?

**Comment by lsf37** [comment]
Looks fine so far. Nitpick: package `jflex.state` should be `jflex.states` to avoid confusion between program state and NFA/DFA states.

**Comment by lsf37** [comment]
Sorry about the conflict, just a comment change.

**Comment by lsf37** [comment]
(merged this to avoid conflicts, hope that was fine)

**Pull request merged**